### ISSUE TITLE
CS8762 false positive when using conditional operator

### ISSUE THREAD

**Issue by thomaslevesque** [dotnet/roslyn#44886]
**Version Used**: 
```
> dotnet --version
5.0.100-preview.4.20258.7
```

**Steps to Reproduce**:

1. Write the following code in a project with nullable reference types enabled
    ```csharp
    using System.Diagnostics.CodeAnalysis;

    namespace TestNullable
    {
        class Test
        {
            static bool TrySomething1(string s, [NotNullWhen(false)] out string? failReason)
            {
                // warning CS8762: Parameter 'failReason' must have a non-null value when exiting with 'false'.
                return s.Length % 2 == 0
                    ? Foo(s, out failReason)
                    : Bar(s, out failReason);
            }

            static bool TrySomething2(string s, [NotNullWhen(false)] out string? failReason)
            {
                // no warning
                if (s.Length % 2 == 0)
                {
                    return Foo(s, out failReason);
                }
                else
                {
                    return Bar(s, out failReason);
                }
            }

            static bool Foo(string s, [NotNullWhen(false)] out string? failReason)
            {
                // Actual implementation not relevant
                failReason = null;
                return true;
            }

            static bool Bar(string s, [NotNullWhen(false)] out string? failReason)
            {
                // Actual implementation not relevant
                failReason = null;
                return true;
            }
        }
    }
    ```
2. Build

**Expected Behavior**:

No warning

**Actual Behavior**:

In `TrySomething1`, the compiler emits this warning:

> warning CS8762: Parameter 'failReason' must have a non-null value when exiting with 'false'.

`TrySomething2` does the same thing with an if/else instead of the conditional operator, and doesn't produce a warning.

**Comment by Youssef1313** [comment]
I can't reproduce with current master. So I think this issue needs a test-only change to be addressed.

**Comment by thomaslevesque** [comment]
@Youssef1313 I confirm that I can no longer reproduce this with after RTM.

**Comment by jcouv** [comment]
Confirmed this no longer repros ([sharplab](https://sharplab.io/#v2:CYLg1APgxAdgrgGwQQwEYIKYAIMzZgWACgABABixIEYA6AEQEtkBzGAewGcAXBgYw5oBhNsAwBBPAgCeHBhwDcxYngC2GDgAdkvbABV1XAHKIU6DMQDexLDcoAmLPu7XbVorY+UqANiyo2bAiOAE5SAMpsalwAFgwwzFQAFNQUHAA0WADahmxGJgDq0biJAGbICBwYAJQAulhscFxeZAD8WGUMCABKGMgcbDBVLp5YbiMjAPQTWADuyMEwccxYgmEAHADs3nYgWAAK88hRGMFYAOQd3b39MGdYKnDcWNHIAG7YyFjsMAC08EhYV7lODYGZFGA4AAeDB48VmMOi5zKFQwZxow3GNhIGywAgAMrhmDEsABSLAOAC8FKwZAxmI8bQAYgFEul6o12shOj0+gMhu56SNdgAheasjINJqXHk3KqKAUjAC+SgVnmovn8gRC4UiGBiSzsySoqQy2VyxiQhWKyMqtXZTRSbWl1z5dNcbsm03Ys3mi3iHs8DBKWFZNAJ8WJZMp1LI/MFNjG8Y82KwzLY4vtnO5LsG8qTytVmIwKIDHkTSaxONFwQzkqzV15udLtgLmILAfVfgCQTTRpNWRyeUt4NK5VtdTrjvrMtdhfdc89WDEvC4cHKWAYKg0mDUMC4yB4Ay+uSwwWLGCBe+bNmdjaw1P+CDz8ZTXGCIOfSpVmM7mqC1b7XFTUHC0ECtGBRxRO1J2NJ0uQbWUA3LekpiXFc1yCTdtwwXd90PCF2CaM9MEvLhr2nHN7y+ExP3pV93wwWiPFbFtiEVIA===)).

We've recently relaxed the enforcement of NotNullWhen attributes in method bodies (maybe in 16.8?). We now only warn when there is a difference in nullability between the WhenTrue and WhenFalse branches at the return statement. Typically, this means when returning a constant (true/false) or returning an invocation that benefits from a conditional attribute.

I'll go ahead and close this as fixed. Thanks
